### PR TITLE
[ AUT-36 ] feat: Enable Exporting Asset stylesheet

### PR DESCRIPTION
**Pull request by bartlomiejmarszal**
This work will enable exporting and importing items containing Shared Stimulus with css stylesheet. 

https://oat-sa.atlassian.net/browse/AUT-36

### CODE REVIEW

File: model/qti/asset/handler/SharedStimulusAssetHandler.php [hunk 1]
```diff
@@ -207,4 +208,17 @@ public function finalize()
...
+
+    private function getSharedStimulusFactory(): SharedStimulusFactory
+    {
+        return $this->getServiceLocator()->get(SharedStimulusFactory::class);
+    }
+
+    private function buildLabelBaseOnParentPath()
+    {
+        $parentPath = $this->getParentPath();
+        $decodedParentPath = json_decode($parentPath);
```
Comment (mike-ionut-mihai-sandu-tao) [inline model/qti/asset/handler/SharedStimulusAssetHandler.php:220]: are you sure it is json in all cases ?

File: model/Export/AbstractQTIItemExporter.php [hunk 2]
```diff
@@ -311,9 +328,48 @@ protected function getServiceManager()
...
+    private function getUniqueAssetDirectoryByLink(MediaBrowser $mediaSource, string $link)
+    {
+        return self::ASSETS_DIRECTORY_NAME
+            . DIRECTORY_SEPARATOR
+            . $mediaSource->getFileInfo($link)['link'];
+    }
+
+    private function buildAssetStylesheetPath(string $basepath, string $baseDirectoryName): string
+    {
+        return $basepath
```
[the remark this answers is not in the document]
Comment (bartlomiejmarszal) [reply model/Export/AbstractQTIItemExporter.php:362]: It's not really readable, isn't it?
Also how would you keep correct order?

File: model/qti/asset/handler/SharedStimulusAssetHandler.php (hunk 1 above)
Comment (bartlomiejmarszal) [reply model/qti/asset/handler/SharedStimulusAssetHandler.php:220]: I will check this.

File: model/Export/AbstractQTIItemExporter.php [hunk 3]
```diff
@@ -311,9 +328,48 @@ protected function getServiceManager()
...
+        if (!$this->metadataExporter) {
             $this->metadataExporter = $this->getServiceManager()->get(MetadataService::SERVICE_ID)->getExporter();
         }
         return $this->metadataExporter;
     }
+
+    /**
+     * @throws AssetStylesheetZipTransferException
+     */
+    private function addAssetStylesheetToZip(string $link, string $baseDirectoryName, string $basepath): void
```
[the remark this answers is not in the document]
Comment (bartlomiejmarszal) [reply model/Export/AbstractQTIItemExporter.php:340]: This is already extracted. What else do you want extract?

File: model/Export/Exception/AssetStylesheetZipTransferException.php [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ * Copyright (c) 2021 (original work) Open Assessment Technologies SA;
+ */
+
+declare(strict_types=1);
+
+namespace oat\taoQtiItem\model\Export\Exception;
+
+use Exception;
+
+class AssetStylesheetZipTransferException extends Exception
```
[the remark this answers is not in the document]
Comment (bartlomiejmarszal) [reply model/Export/Exception/AssetStylesheetZipTransferException.php:27]: We should keep high granularity of exceptions and if possible group those exceptions into groups that are easy to catch and handle. What I did is that I added service exception and I added catch block for it.

File: model/Export/Stylesheet/AssetStylesheetLoader.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+{
+    use OntologyAwareTrait;
+
+    public const PROPERTY_LINK = 'http://www.tao.lu/Ontologies/TAOMedia.rdf#Link';
+    public const ASSET_CSS_DIRECTORY_NAME = 'css';
+    public const ASSET_CSS_FILENAME = 'tao-user-styles.css';
+
+    public function loadAssetFromAssetResource(string $link): ?StreamInterface
+    {
+        $asset = $this->getResource(UriHelper::decode($link));
```
[the remark this answers is not in the document]
Comment (bartlomiejmarszal) [reply model/Export/Stylesheet/AssetStylesheetLoader.php:41]: IMO we should not avoid static helpers for all cost. Those methods are fast and useful and as long as they are not changing a state of class (so i.e string operations) we should use them.

File: model/qti/asset/handler/SharedStimulusAssetHandler.php [hunk 6]
```diff
@@ -84,27 +92,20 @@ public function isApplicable($relativePath)
      */
     public function handle($absolutePath, $relativePath)
     {
-        $sharedFiles = $this->getSharedFiles();
-
-        $md5 = md5_file($absolutePath);
```
[the remark this answers is not in the document]
Comment (bartlomiejmarszal) [reply model/qti/asset/handler/SharedStimulusAssetHandler.php:89]: When I read this method before change I understood that this md5 was created as a key for shared file that we store for no reason. I did not found a good reason behind it.

File: model/Export/Stylesheet/AssetStylesheetLoader.php [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+                $this->buildAssetPathFromPropertyName($property)
+            );
+        }
+
+        return null;
+    }
+
+    private function buildAssetPathFromPropertyName(string $property)
+    {
+        return dirname($property)
```
[the remark this answers is not in the document]
Comment (bartlomiejmarszal) [reply model/Export/Stylesheet/AssetStylesheetLoader.php:58]: Where is this duplicated? It's relative path building here. I don't think I can find this in AbstractQTIItemExporter.

File: model/qti/asset/handler/SharedStimulusAssetHandler.php (hunk 1 above)
Comment (bartlomiejmarszal) [reply model/qti/asset/handler/SharedStimulusAssetHandler.php:220]: I found only one usage of this attribute in `\oat\taoQtiItem\model\qti\ImportService::importQtiItem` 
For some reason we encode and set this attribute as json. I adjust SharedStimulusHandler to existing code.

**Review by pnal — COMMENTED** [review]
Good job!

File: model/qti/asset/handler/SharedStimulusAssetHandler.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (bartlomiejmarszal) [reply model/qti/asset/handler/SharedStimulusAssetHandler.php:89]: This is very old code here. I'm not using attribute `$sharedFiles` at all. I had to change a way how we manipulate files and I extracted this to external service. This is not causing any errors so far. I call this improvement in this method. Can we check this on QA if it's causing any errors?

[the remark this answers is not in the document]
Comment (mike-ionut-mihai-sandu-tao) [reply model/qti/asset/handler/SharedStimulusAssetHandler.php:89]: as long as we have this inside the integration branch and we will do a regression before merging then yes

File: model/Export/AbstractQTIItemExporter.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (mike-ionut-mihai-sandu-tao) [reply model/Export/AbstractQTIItemExporter.php:340]: i wanted to extract this Css functionality inside a different service as the AbstractQTIItemExporter.php should be like a root agregator

**Review by mike-ionut-mihai-sandu-tao — APPROVED** [review]
because this will go inside a integration branch and because we will have to refactor some bits i am ok with the current working solution 
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful

**Review by pnal — APPROVED** [review]
- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful

**Pull request merged**